src/Database/DatabaseConfig.php

```diff
@@ -114,6 +120,43 @@ public function connection(): array
         );
     }
 
+    /**
+     * Tenant's host database connection config.
+     */
+    public function hostConnection(): array
+    {
+        $config = $this->tenantConfig();
+        $template = $this->getTemplateConnectionName();
+        $templateConnection = config("database.connections.{$template}");
+
+        if ($this->manager() instanceof Contracts\ManagesDatabaseUsers) {
+            // We don't need username and password for database creation/deletion
+            // Username and password will be saved in Tenant's config
+            // and will be used for tenancy db connection
+            unset($config['username']);
+            unset($config['password']);
+        }
```

Comment:
Is this code needed, or does it just simplify the `$config` array?

From the `We don't need username` it feels as if you were just simplifying the data, but I assume there's a better reason for why this is being unset?

Reply:
> Is this code needed, or does it just simplify the $config array?

This code is needed because if we don't remove the username and password from the config, an error occurred while establishing a connection like `SQLSTATE[HY000] [1045] Access denied for user '3qVdsacitd420N25'@'172.24.0.7' (using password: YES) (SQL: SELECT SCHEMA_NAME FROM INFORMATION_SCHEMA.SCHEMATA WHERE SCHEMA_NAME = 'foo1k7L3u4g')`. 
An error occurred only for the tests using the `PermissionControlledMySQLDatabaseManager`. 

I'm trying but unable to come up with better comments.

Reply:
I see, so the problem is that we don't want to use the tenant's DB credentials since they don't have the permissions to create databases? And I assume these values (`username` and `password`) get filled with the "main" credentials? Where does that happen?

> I'm trying but unable to come up with better comments.

It should just say why we're doing this, not that "We don't need" the username. Not needing the username is different from *having* to remove the username for this to work.

Reply:
> And I assume these values (username and password) get filled with the "main" credentials? Where does that happen?

It got filled in `makeCredentials()` method. https://github.com/archtechx/tenancy/blob/f97651e16ba300c5e4671031a19c101f1f0fae42/src/Database/DatabaseConfig.php#L88-L91
And this method got called in `CreateDatabase` job. https://github.com/archtechx/tenancy/blob/cd6e989c09981800aef439471d56e9654b1bbbf1/src/Jobs/CreateDatabase.php#L36

Reply:
Okay, so this set the username and password that you're later unsetting? But could you expand on what I asked?

> And I assume these values (username and password) get filled with the "main" credentials? Where does that happen?

I'm asking about what happens **after** you unset these values, not before.

Reply:
Ah Ok. 

> values (username and password) get filled with the "main" credentials? Where does that happen?

We are removing the username and password from the **tenant's config**, and the template connection still has "main" credentials. We are doing an array replace thing at the end and that's how main credentials got filled. 
https://github.com/archtechx/tenancy/blob/7f6166d49a6b07f8b4c4831f12a6ddc15c305b88/src/Database/DatabaseConfig.php#L143

Reply:
> and the template connection still has "main" credentials. We are doing an array replace thing at the end and that's how main credentials got filled.

Okay, so the template connection always has credentials, and what you're doing in the `unset()` is removing credentials **from the config** that later gets merged into the template connection? That way we prevent the credentials in the config from overriding the template connection credentials?

And where do the credentials for `$templateConnection` come from?

Reply:
> And where do the credentials for $templateConnection come from?

it is coming from the `config("database.connections.{$template}");` https://github.com/archtechx/tenancy/blob/7f6166d49a6b07f8b4c4831f12a6ddc15c305b88/src/Database/DatabaseConfig.php#L130-L131. `$template` generated from another method https://github.com/archtechx/tenancy/blob/7f6166d49a6b07f8b4c4831f12a6ddc15c305b88/src/Database/DatabaseConfig.php#L98-L103